src/kernels/webgl/mulmat_packed_gpu_cs_v4.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {GPGPUProgram} from './gpgpu_math';
+
+export class MatMulPackedProgramCSV4 implements GPGPUProgram {
+  variableNames = ['matrixA', 'matrixB'];
+  usesPackedTextures = true;
+  outputShape: number[];
+  userCode: string;
+  localGroupSize: number[];
+  workPerThread: number[];
+
+  constructor(
+      aShape: [number, number, number], outputShape: [number, number, number],
+      transposeA = false, transposeB = false, addBias = false,
+      activation: string = null) {
+    this.outputShape = outputShape;
+
+    const sharedDim = transposeA ? aShape[1] : aShape[2];
+    const sharedDimensionPacked = Math.ceil(sharedDim / 2);
+
+    const TSM = 64, TSN = 64, TSK = 16;
+    const WPTM = 4, WPTN = 4;
+    const LPTA = TSK * WPTM * WPTN / TSN;
+    // const LPTB = TSK * WPTM * WPTN / TSM;
+    const RTSM = 16, RTSN = 16;
+
+    this.localGroupSize = [RTSN, RTSM];
+    this.workPerThread = [WPTN, WPTM];
+
+    const aSample = transposeA ? `tiledIndex * 2, (offsetM + row) * 2` :
+                                 `(offsetM + row) * 2, tiledIndex * 2`;
+    const bSample = transposeB ? `(offsetN + row) * 2, tiledIndex * 2` :
+                                 `tiledIndex * 2, (offsetN + row) * 2`;
+    const aSwizzle = transposeA ? ['a.xxyy', 'a.zzww'] : ['a.xxzz', 'a.yyww'];
+    const bSwizzle = transposeB ? ['b.xzxz', 'b.ywyw'] : ['b.xyxy', 'b.zwzw'];
+
+    let activationSnippet = '', applyActivationSnippet = '';
+    if (activation) {
+      activationSnippet = `vec4 activation(vec4 x) {
+        ${activation}
+      }`;
+      applyActivationSnippet = `acc[wm][wn] = activation(acc[wm][wn]);`;
+    }
+
+    if (addBias) {
+      console.error('bias is not supported');
+    }
+
+    this.userCode = `
+      ${activationSnippet}
+
+      shared vec4 Asub[${TSM}][${TSK}];
+      shared vec4 Bsub[${TSK}][${TSN}];
+
```

Comment:
I tried `Asub[TSM][TSK+2],Asub[TSM+2][TSK],Bsub[TSK][TSN+2],Bsub[TSK+2][TSN]`.
It seems performance has not changed.

Reply:
In the texture version, we use RGBA 4 channels which is equal to vec4. But in https://www.ibiblio.org/e-notes/webgl/gpu/mul/sgemm6ib.htm, it uses float not vec4. So we should reduce WPTM/WPTN or RTSM/RTSN to see if the performance will be impacted?

Reply:
ok i will register flags and test.